### ISSUE TITLE
non-blocking cancellation

### ISSUE THREAD

**Issue by dlax** [psycopg/psycopg#534]
Postgres ~~16~~ 17 might include a libpq API for non-blocking cancellation: https://commitfest.postgresql.org/37/3511/

This seems interesting to me, and I wonder if this would help with respect to various issues related to cancellations that occurred some times ago (ctrl+c, etc.).

I have prepared a ~~["pg16/non-blocking-pqcancel" branch](https://github.com/dlax/psycopg3/tree/pg16/non-blocking-pqcancel)~~ ["pg17/non-blocking-pqcancel" branch](https://github.com/dlax/psycopg3/tree/pg17/non-blocking-pqcancel) on my fork to include this (uncommitted) work into Psycopg; the result is that `AsyncConnection.cancel()` becomes `async`, so it's a breaking change.

Obviously, we'd wait until this eventually lands in Postgres ~~16~~ 17. I just wanted to share this work in some way and possibly discuss the idea a bit.

**Comment by dvarrazzo** [comment]
Hey, I just happened to read your message on -hackers, and that's great! I also took a look at your code.

It would be sad to introduce this change only in psycopg 4. It would be better to add an `async AsyncConnection.acancel()` function on the Async connection, and to make `AsyncConnection.cancel()` raise a warning when used. We could release it in a future psycopg 3.x.

Before PG 16 is released (or in case this libpq feature doesn't make the cut), we could actually implement `acancel()` calling `PQcancel()` in a thread executor and release at least this in psycopg 3.2, so that the interface should be in place.

**Comment by dlax** [comment]
> It would be sad to introduce this change only in psycopg 4. It would be better to add an `async AsyncConnection.acancel()` function on the Async connection, and to make `AsyncConnection.cancel()` raise a warning when used. We could release it in a future psycopg 3.x.

Yes, we can do that, sure. But, in fact, maybe we can keep `AsyncConnection.cancel()` implemented as is it now so as to keep something that can be used within a signal handler (the new libpq API is not supposed to be used in that case)?

**Comment by dvarrazzo** [comment]
> maybe we can keep AsyncConnection.cancel() implemented as is it now so as to keep something that can be used within a signal handler

Sure, I didn't think about this use case. We can document that `cancel()` is blocking on `AsyncConnection` and add a `acancel()` method on a future feature release.

**Comment by dlax** [comment]
Patches have been committed in PostgreSQL; so I'm resuming my prior work for integration in Psycopg.

**Comment by dlax** [comment]
For future reference, the commit in PostgreSQL is [61461a300c1cb5d53955ecd792ad0ce75a104736](https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=61461a300c1cb5d53955ecd792ad0ce75a104736) "libpq: Add encrypted and non-blocking query cancellation routines". In addition to the original intent of making cancellation non-blocking, it also allows it to be encrypted. As a consequence, the PostgreSQL documentation now states that the old API:
> [PQcancel](https://www.postgresql.org/docs/devel/libpq-cancel.html#LIBPQ-PQCANCEL) is a deprecated and insecure variant of [PQcancelBlocking](https://www.postgresql.org/docs/devel/libpq-cancel.html#LIBPQ-PQCANCELBLOCKING), but one that can be used safely from within a signal handler.

The issue in Psycopg is that `AsyncConnection.cancel()` is not an async function, so we cannot directly use this new libpq API there. Accordingly, following prior discussion here, we'd introduce an `acancel()` async method to use this new API.

For the sync `Connection`, the `cancel()` method could also benefit from the new non-blocking API (probably more its "secure" feature) but if we want to preserve the "signal handler" use case, we need to handle two implementations. For this, I would suggest to add an `insecure: bool | None = None` parameter to `cancel()`; then:
- `cancel(insecure=False)` would use the new non-blocking and secure cancellation API
- `cancel(insecure=True)` would call `PQcancel` explicitly for the signal handler use case
- by defaulting to `insecure=None`, meaning use the old `PQcancel` API, we preserve backwards-compatibility

Open questions:
1. Should `AsyncConnection.acancel()` raise a `NotSupportedError` if the libpq is too old or fall back to the old `PQcancel` API?
2. Should we warn about calling `Connection.cancel()` without an explicit `bool` `insecure` argument (the reason for `insecure: bool | None = None`)? In preparation for changing `insecure: bool = False` for Psycopg 4?
3. Or are all proposed changes to `Connection.cancel()` not worth after all? (or not now)
4. Whenever we internally use `cancel()` (e.g. when handling `KeyboardInterrupt` in `*Connection.wait()` or in `*Cursor.stream()`), should we use the non-blocking API? (Seems an harmless improvement to me.)

**Comment by dvarrazzo** [comment]
Hello @dlax, thank you for this massive work!

> 1. Should AsyncConnection.acancel() raise a NotSupportedError if the libpq is too old or fall back to the old PQcancel API?

I think that raising an exception would be impolite... Probably we can run `PQcancel()` in a thread executor in order to maintain the non-blocking promise.

> 2. Should we warn about calling Connection.cancel() without an explicit bool insecure argument (the reason for insecure: bool | None = None)? In preparation for changing insecure: bool = False for Psycopg 4?
>
> 3. Or are all proposed changes to Connection.cancel() not worth after all? (or not now)

I think they are worth, but I think that the combination of libpq version not widely supported, desire of not breaking the current interface, make changing the `cancel()` signature pretty awkward.

What about, instead of introducing `acancel()`, to introduce a `Connection.cancel_safe()` method, which would be:

- async on async connection
- raise NotSupportedError if libpq < 17, or, I think better,
- call legacy `cancel()` if libpq < 17 (in an executor if called from async) [see later for the rationale].

We could leave the `cancel()` method with its limitation as it is. I would even avoid to raise a warning if the old `cancel()` is called because it's difficult to tell apart if it's been called from a handler or not. I would rather document well its limitations. As seen in #753, calling cancel from a signal is pretty useful.

I've read the current state of #763, and I have a couple of comments. I'll avoid to comment there because maybe that implementation might change considerably if we go for the `cancel_safe()` option:

- how about calling `pgconn.get_cancel()` from within `cancel_gen()`?
- how about adding a method `psycopg.pq.PGcancelConn.is_supported()` to allow people (and ourselves) to check if we can use the newer methods?
- how about adding a timeout to `cancel_safe()`? This would be better defined after merging #766.

Why would I make `cancel_safe()` call `cancel()` as fallback? Because everyone should do it, [including](https://github.com/psycopg/psycopg/pull/763/files#diff-b3b8cfbbb6a3720571b3ceceb3dc998d151667f62e5738a225bd0234cbde51ffR64-R67) [yourself](https://github.com/psycopg/psycopg/pull/763/files#diff-d1ce6edd42dfb409b798d888a60cdb0a7ee7db036ce6feb45a4b1ebdf318b32dR163-R168). It would allow to embed a cancel in the code flow more easily, implementing your point 4 easily and making it even easier for async_to_sync.

> 4. Whenever we internally use cancel() (e.g. when handling KeyboardInterrupt in *Connection.wait() or in *Cursor.stream()), should we use the non-blocking API? (Seems an harmless improvement to me.)

Yes, I would make these changes, which I think would be simple to make if we have an `[async] cancel_safe()` with fallback.

**Comment by dvarrazzo** [comment]
> - how about adding a method psycopg.pq.PGcancelConn.is_supported() to allow people (and ourselves) to check if we can use the newer methods?

The need for a public method is probably not huge if `cancel_safe()` uses the callback, as the method would be always available.

**Comment by dlax** [comment]
Thank you for your inputs.
I'll wait for #754 to get in before getting back to the high-level interface in #763.

**Comment by dlax** [comment]
Ok; I've implemented most of your suggestions in #763 now, leaving appart:
- `PGcancelConn.is_supported()`, as suggested above
- the timeout parameter to `cancel_safe()`, as it seems to me that this could be handled by the end-user as:
   ```python
   async with asyncio.timeout(10):
       await conn.cancel_safe()
   ```

**Comment by dvarrazzo** [comment]
Let's leave is_supported out, yes, especially if we introduce #772.

I'll take a look to the rest, thank you very much :slightly_smiling_face:
